README.md
sml-serial-receiver
===================

jSML is a Java library implementing the Smart Message Language (SML).

This repository contains a small patch to the jSML SerialRead example program
that provides a more human-readable representation of the data read from
electronic energy meters (specifically the EMH ED300L).

Licensing
---------

jSML v1.0.17 was released under the GNU Lesser General Public License v2.1 (see
[licenses/COPYING.LESSER]) thus requiring all modifications to be licensed
under the LGPL v2.1 as well.

Note that as of v1.1.0 (19-Jul-2017), jSML is licensed under the Mozilla Public
License v2.0.


doc/CHANGELOG.txt
v1.0.17 19-Jun-2014
-------------------
- changed the license from LGPL version 3 to LGPL version 2.1
- library is uploaded to central repository
- fixed bug in Integer64 using large integers (>32 bit) (thanks to
  Patrick Hinzke from Kellendonk)
- closed streams correctly in SML_TConnection (thanks to Patrick
  Hinzke from Kellendonk)
- fixed bug in List_of_SML_ObjReqEntry (thanks to Patrick
  Hinzke from Kellendonk)

v1.0.16 30-Dec-2013
-------------------
- a syntax check of the tl field in SequenceOf was removed so that
  jSML is compatible with Hager EDL meters. Thanks to Jochen Burkhardt

v1.0.15 02-Oct-2013
-------------------
- small bug fix in receive function of SML_TConnection thanks to David
  Croft

v1.0.14 10-Feb-2012
-------------------
- another bug fix in decodeAndCheck of SML_Message.java thanks to
  Roland Schroth from RWE

v1.0.13 06-Feb-2012
-------------------
- no code changes, only minor changes to the gradle build file

v1.0.12 13-Jan-2012
-------------------
- addition of SML_TimestampLocal from the SML 1.4 draft thanks to
  Prosyst

v1.0.11 29-Nov-2011
-------------------
- connectTo Function can be passed a timeout. Thanks to Fabian
  Kohlmorgen from FH-Dortmund for the idea.
- CRC of SML_Message is now only checked if decodeAndCheck function is
  called
- switched to the Gradle build system

v1.0.10 07-Nov-2011
-------------------
- another bug fix in SML_Message.java when checking CRC thanks to
  Fabian Kohlmorgen from FH-Dortmund and prosyst
- a small fix in OctetString.hashCode() thanks to Roland Schroth from
  RWE

v1.0.9 20-Oct-2011
------------------
- bug fix in SML_Message.java when checking CRC thanks to Fabian
  Kohlmorgen from FH-Dortmund

v1.0.8 05-Aug-2011
------------------
- bug fix in Unsigned64.java, thanks to Jan Heise and Boris Pigin from
  Vattenfall

v1.0.7 06-Apr-2011
------------------
- fixed bug in SML_ProcParValue
- changed to maven build system

v1.0.6 15-Feb-2011
------------------
- fixed synchronization problem in SML_SerialReceiver

v1.0.5 20-Jan-2011
------------------
- improved the SSL support and included a sample that demonstrates the
  use of SSL

v1.0.3 16-Dec-2010
------------------
- added a simple class to receive SMLMessages (SML_SerialReceiver)
  over a serial interface
- This class can be used to read out the eHZ Meter

v1.0.2 27-Aug-2010
------------------
- Additions to the OctetString Class - Thanks to ProSyst for the
  contribution

v1.0.1 15-Jul-2010
------------------
- SSL Support for the Server.
- Added some missing SML-Objects

v1.0 04-Mai-2010
----------------
- completely restructured the SML Transport Layer implementation and
  API!
- removed all loggers

v0.9.1 18-Dec-2009
------------------
- Fixed CRC
- fixed bug in ImplicitChoice


src/test/java/org/openmuc/jsml/tl/ClientServerTCPITest.java
/*
 * Copyright 2009-14 Fraunhofer ISE
 *
 * This file is part of jSML.
 * For more information visit http://www.openmuc.org
 *
 * jSML is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * jSML is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with jSML.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.openmuc.jsml.tl;

import static org.junit.Assert.assertTrue;

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;

import org.junit.Before;
import org.junit.Test;
import org.openmuc.jsml.structures.Integer32;
import org.openmuc.jsml.structures.OctetString;
import org.openmuc.jsml.structures.SML_GetListReq;
import org.openmuc.jsml.structures.SML_GetListRes;
import org.openmuc.jsml.structures.SML_List;
import org.openmuc.jsml.structures.SML_ListEntry;
import org.openmuc.jsml.structures.SML_Message;
import org.openmuc.jsml.structures.SML_MessageBody;
import org.openmuc.jsml.structures.SML_PublicCloseReq;
import org.openmuc.jsml.structures.SML_PublicOpenReq;
import org.openmuc.jsml.structures.SML_PublicOpenRes;
import org.openmuc.jsml.structures.SML_Status;
import org.openmuc.jsml.structures.SML_Unit;
import org.openmuc.jsml.structures.SML_Value;
import org.openmuc.jsml.structures.Unsigned8;

public class ClientServerTCPITest implements SML_TConnectionListener {

	public final static int serverPort = 53214;

	@Before
	public void startTCPServer() throws IOException {
		SML_TSAP sml_tSAP = null;

		sml_tSAP = new SML_TSAP(serverPort, this);

		sml_tSAP.startListening();

	}

	@Test
	public void testClient() throws UnknownHostException, IOException {

		/* Create client TSAP */
		SML_TSAP sml_tSAP = new SML_TSAP();

		SML_TConnection sml_tConnection = sml_tSAP.connectTo(InetAddress.getByName("localhost"), serverPort, 0);

		byte[] getListRequest = createGetListRequest(0, 1);
		// Assert.assertTrue("getListRequest not coded correctly",
		// Arrays.equals(getListRequest, getListRequestCheck));
		sml_tConnection.send(getListRequest);
		sml_tConnection.receive();
	}

	/************************************
	 * Client
	 */

	/**
	 * This sample builds a SML-File containing a PublicOpenRequest, a GetListRequest and a PublicCloseRequest
	 * 
	 * @return
	 */
	public static byte[] createGetListRequest(int groupNumber, int transactionIDCounter) {

		ByteArrayOutputStream bs = new ByteArrayOutputStream(50);
		DataOutputStream os = new DataOutputStream(bs);

		// --- create a SML_Message with SML_PublicOpenRequest to start the
		// SML_File
		// make the transacionId and increment the transactionIDCounter
		OctetString transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		// make the group number and increment the group number
		Unsigned8 groupNum = new Unsigned8(groupNumber);
		// continue on error
		Unsigned8 abortOnError = new Unsigned8(0x00);
		// create clientId
		int iClientId = 0x1;
		OctetString clientId = new OctetString(("" + iClientId).getBytes());
		// create reqFileId
		int iReqFileId = 0x11;
		OctetString reqFileId = new OctetString(("" + iReqFileId).getBytes());
		// create the SML_Message
		SML_PublicOpenReq openRequest = new SML_PublicOpenReq(null, clientId, reqFileId, null, null, null, null);
		SML_MessageBody smlMessageBody = new SML_MessageBody(SML_MessageBody.OpenRequest, openRequest);
		SML_Message openRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// --- create a SML_Message with SML_GetProfileListReq with test values
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		OctetString listName = new OctetString(new byte[] { 0x01, 0x00, 0x01, 0x08, 0x01, (byte) 0xff });
		SML_GetListReq getListRequest = new SML_GetListReq(clientId, null, null, null, listName);
		smlMessageBody = new SML_MessageBody(SML_MessageBody.GetListRequest, getListRequest);
		SML_Message getListRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// --- create a SML_Message with SML_PublicCloseRequest to end the
		// SML_File
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		SML_PublicCloseReq closeRequest = new SML_PublicCloseReq(null);
		smlMessageBody = new SML_MessageBody(SML_MessageBody.CloseRequest, closeRequest);
		SML_Message closeRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// encode every SML_Messages and write the encoded data to the stream os
		try {
			openRequestMessage.code(os);
			getListRequestMessage.code(os);
			closeRequestMessage.code(os);
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}

		return bs.toByteArray();
	}

	/**
	 * This sample builds a SML-File containing a PublicOpenResponse, a GetListResponse and a PublicCloseResponse
	 * 
	 * @return
	 */
	public static byte[] createGetListResponse(int groupNumber, int transactionIDCounter) {

		ByteArrayOutputStream bs = new ByteArrayOutputStream(50);
		DataOutputStream os = new DataOutputStream(bs);

		// --- create a SML_Message with SML_PublicOpenRequest to start the
		// SML_File
		// make the transacionId and increment the transactionIDCounter
		OctetString transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		// make the group number and increment the group number
		Unsigned8 groupNum = new Unsigned8(groupNumber);
		// continue on error
		Unsigned8 abortOnError = new Unsigned8(0x00);
		// create clientId
		int iServerId = 0x1;
		OctetString serverId = new OctetString(("" + iServerId).getBytes());
		// create reqFileId
		int iReqFileId = 0x11;
		OctetString reqFileId = new OctetString(("" + iReqFileId).getBytes());
		// create the SML_Message
		SML_PublicOpenRes openResponse = new SML_PublicOpenRes(null, null, reqFileId, serverId, null, null);
		SML_MessageBody smlMessageBody = new SML_MessageBody(SML_MessageBody.OpenResponse, openResponse);
		SML_Message openRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// --- create a SML_Message with SML_GetProfileListReq with test values
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());

		new SML_Status(new Unsigned8(0));
		SML_Unit unit = new SML_Unit(new Unsigned8(30));
		SML_Value value = new SML_Value(new Integer32(0));

		SML_ListEntry[] valListEntries = new SML_ListEntry[1];
		valListEntries[0] = new SML_ListEntry(
				new OctetString(new byte[] { 0x01, 0x00, 0x01, 0x08, 0x01, (byte) 0xff }), null, null, unit, null,
				value, null);
		SML_List valList = new SML_List(valListEntries);
		SML_GetListRes getListResponse = new SML_GetListRes(null, serverId, null, null, valList, null, null);

		smlMessageBody = new SML_MessageBody(SML_MessageBody.GetListResponse, getListResponse);
		SML_Message getListRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// --- create a SML_Message with SML_PublicCloseRequest to end the
		// SML_File
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		SML_PublicCloseReq closeRequest = new SML_PublicCloseReq(null);
		smlMessageBody = new SML_MessageBody(SML_MessageBody.CloseResponse, closeRequest);
		SML_Message closeRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// encode every SML_Messages and write the encoded data to the stream os
		try {
			openRequestMessage.code(os);
			getListRequestMessage.code(os);
			closeRequestMessage.code(os);
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}

		return bs.toByteArray();
	}

	/************************************
	 * Server
	 * 
	 * @throws IOException
	 */

	@Override
	public void connectionIndication(SML_TConnection sml_tConnection) {
		try {
			sml_tConnection.receive();
			byte[] getListResponse = createGetListResponse(0, 1);
			sml_tConnection.send(getListResponse);
		} catch (IOException e) {
			e.printStackTrace();
			assertTrue(false);
		}

	}

	@Override
	public void serverStoppedListeningIndication(IOException e) {
		// TODO Auto-generated method stub

	}

}


src/test/java/org/openmuc/jsml/structures/DecodingEDL40Test.java
/*
 * Copyright 2009-14 Fraunhofer ISE
 *
 * This file is part of jSML.
 * For more information visit http://www.openmuc.org
 *
 * jSML is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * jSML is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with jSML.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.openmuc.jsml.structures;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;

import org.junit.Assert;
import org.junit.Test;

public class DecodingEDL40Test {

	@Test
	public void decodeTimeTagMessage() throws IOException {
		byte[] message = new byte[] { (byte) 0x76, (byte) 0x09, (byte) 0x67, (byte) 0x08, (byte) 0x00, (byte) 0x00,
				(byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x62, (byte) 0x04, (byte) 0x62, (byte) 0x00,
				(byte) 0x72, (byte) 0x63, (byte) 0x01, (byte) 0x01, (byte) 0x76, (byte) 0x01, (byte) 0x07, (byte) 0x1f,
				(byte) 0x15, (byte) 0xb0, (byte) 0x8c, (byte) 0x68, (byte) 0x5c, (byte) 0x09, (byte) 0xcd, (byte) 0x02,
				(byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x09, (byte) 0x08,
				(byte) 0x0c, (byte) 0x2a, (byte) 0xec, (byte) 0x2d, (byte) 0x4c, (byte) 0x65, (byte) 0x63, (byte) 0x01,
				(byte) 0x01, (byte) 0x63, (byte) 0x6c, (byte) 0xc5, (byte) 0x00, (byte) 0x76, (byte) 0x09, (byte) 0x68,
				(byte) 0x08, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x62,
				(byte) 0x05, (byte) 0x62, (byte) 0x00, (byte) 0x72, (byte) 0x63, (byte) 0x07, (byte) 0x01, (byte) 0x77,
				(byte) 0x01, (byte) 0x09, (byte) 0x08, (byte) 0x0c, (byte) 0x2a, (byte) 0xec, (byte) 0x2d, (byte) 0x4c,
				(byte) 0x65, (byte) 0x63, (byte) 0x01, (byte) 0x72, (byte) 0x62, (byte) 0x01, (byte) 0x64, (byte) 0x1b,
				(byte) 0xcb, (byte) 0x2d, (byte) 0x7e, (byte) 0x77, (byte) 0x07, (byte) 0x81, (byte) 0x81, (byte) 0xc7,
				(byte) 0x82, (byte) 0x03, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x04,
				(byte) 0x45, (byte) 0x4d, (byte) 0x48, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00,
				(byte) 0x00, (byte) 0x00, (byte) 0x09, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01,
				(byte) 0x09, (byte) 0x08, (byte) 0x0c, (byte) 0x2a, (byte) 0xec, (byte) 0x2d, (byte) 0x4c, (byte) 0x65,
				(byte) 0x63, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x01, (byte) 0x08,
				(byte) 0x00, (byte) 0xff, (byte) 0x63, (byte) 0x01, (byte) 0x88, (byte) 0x01, (byte) 0x62, (byte) 0x1e,
				(byte) 0x52, (byte) 0xff, (byte) 0x54, (byte) 0x03, (byte) 0xe9, (byte) 0x1f, (byte) 0x01, (byte) 0x77,
				(byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x02, (byte) 0x08, (byte) 0x00, (byte) 0xff, (byte) 0x63,
				(byte) 0x01, (byte) 0x88, (byte) 0x01, (byte) 0x62, (byte) 0x1e, (byte) 0x52, (byte) 0xff, (byte) 0x53,
				(byte) 0x0a, (byte) 0x61, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x01,
				(byte) 0x08, (byte) 0x01, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x62, (byte) 0x1e, (byte) 0x52,
				(byte) 0xff, (byte) 0x54, (byte) 0x03, (byte) 0xe9, (byte) 0x1f, (byte) 0x01, (byte) 0x77, (byte) 0x07,
				(byte) 0x01, (byte) 0x00, (byte) 0x02, (byte) 0x08, (byte) 0x01, (byte) 0xff, (byte) 0x01, (byte) 0x01,
				(byte) 0x62, (byte) 0x1e, (byte) 0x52, (byte) 0xff, (byte) 0x53, (byte) 0x0a, (byte) 0x61, (byte) 0x01,
				(byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x01, (byte) 0x08, (byte) 0x02, (byte) 0xff,
				(byte) 0x01, (byte) 0x01, (byte) 0x62, (byte) 0x1e, (byte) 0x52, (byte) 0xff, (byte) 0x52, (byte) 0x00,
				(byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x0f, (byte) 0x07, (byte) 0x00,
				(byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x62, (byte) 0x1b, (byte) 0x52, (byte) 0xff, (byte) 0x53,
				(byte) 0x02, (byte) 0x4c, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x23,
				(byte) 0x07, (byte) 0x00, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x62, (byte) 0x1b, (byte) 0x52,
				(byte) 0xff, (byte) 0x52, (byte) 0x49, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00,
				(byte) 0x37, (byte) 0x07, (byte) 0x00, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x62, (byte) 0x1b,
				(byte) 0x52, (byte) 0xff, (byte) 0x53, (byte) 0x00, (byte) 0x94, (byte) 0x01, (byte) 0x77, (byte) 0x07,
				(byte) 0x01, (byte) 0x00, (byte) 0x4b, (byte) 0x07, (byte) 0x00, (byte) 0xff, (byte) 0x01, (byte) 0x01,
				(byte) 0x62, (byte) 0x1b, (byte) 0x52, (byte) 0xff, (byte) 0x53, (byte) 0x01, (byte) 0x6f, (byte) 0x01,
				(byte) 0x77, (byte) 0x07, (byte) 0x01, (byte) 0x00, (byte) 0x01, (byte) 0x11, (byte) 0x00, (byte) 0xff,
				(byte) 0x63, (byte) 0x01, (byte) 0x88, (byte) 0x72, (byte) 0x62, (byte) 0x03, (byte) 0x73, (byte) 0x65,
				(byte) 0x4e, (byte) 0xdf, (byte) 0x67, (byte) 0x4c, (byte) 0x52, (byte) 0x00, (byte) 0x52, (byte) 0x00,
				(byte) 0x62, (byte) 0x1e, (byte) 0x52, (byte) 0xff, (byte) 0x54, (byte) 0x03, (byte) 0xe9, (byte) 0x12,
				(byte) 0x83, (byte) 0x04, (byte) 0x0f, (byte) 0xdf, (byte) 0xa7, (byte) 0xe7, (byte) 0xd8, (byte) 0x0e,
				(byte) 0xfb, (byte) 0xf5, (byte) 0xf0, (byte) 0xcf, (byte) 0xe8, (byte) 0x70, (byte) 0x5a, (byte) 0x43,
				(byte) 0x10, (byte) 0x23, (byte) 0x52, (byte) 0xb8, (byte) 0x92, (byte) 0x68, (byte) 0x22, (byte) 0xde,
				(byte) 0x02, (byte) 0x37, (byte) 0x0b, (byte) 0xb0, (byte) 0xac, (byte) 0x19, (byte) 0x0e, (byte) 0xcc,
				(byte) 0x90, (byte) 0x41, (byte) 0xce, (byte) 0x2f, (byte) 0xc0, (byte) 0x12, (byte) 0x27, (byte) 0x6a,
				(byte) 0x28, (byte) 0xc9, (byte) 0xf3, (byte) 0x73, (byte) 0xfb, (byte) 0x1f, (byte) 0xb9, (byte) 0x68,
				(byte) 0x38, (byte) 0x7f, (byte) 0x00, (byte) 0x03, (byte) 0x77, (byte) 0x07, (byte) 0x81, (byte) 0x81,
				(byte) 0xc7, (byte) 0x82, (byte) 0x05, (byte) 0xff, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01,
				(byte) 0x83, (byte) 0x02, (byte) 0x60, (byte) 0xd6, (byte) 0x6c, (byte) 0x2e, (byte) 0xf2, (byte) 0x3e,
				(byte) 0xdf, (byte) 0x14, (byte) 0xc7, (byte) 0x60, (byte) 0x62, (byte) 0x23, (byte) 0x2e, (byte) 0x3f,
				(byte) 0x6e, (byte) 0xc5, (byte) 0x85, (byte) 0x96, (byte) 0x6c, (byte) 0x4b, (byte) 0x40, (byte) 0x0c,
				(byte) 0x39, (byte) 0x0d, (byte) 0xa7, (byte) 0xf0, (byte) 0xf4, (byte) 0xf4, (byte) 0x47, (byte) 0x35,
				(byte) 0x45, (byte) 0x08, (byte) 0x5c, (byte) 0xc8, (byte) 0x25, (byte) 0x29, (byte) 0x37, (byte) 0xad,
				(byte) 0x92, (byte) 0x82, (byte) 0xd7, (byte) 0x97, (byte) 0x6c, (byte) 0xda, (byte) 0x0a, (byte) 0x78,
				(byte) 0x51, (byte) 0x4c, (byte) 0x01, (byte) 0x77, (byte) 0x07, (byte) 0x99, (byte) 0x00, (byte) 0x00,
				(byte) 0x00, (byte) 0x00, (byte) 0x11, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x69,
				(byte) 0x00, (byte) 0x00, (byte) 0x01, (byte) 0x00, (byte) 0x01, (byte) 0x08, (byte) 0x02, (byte) 0xff,
				(byte) 0x01, (byte) 0x01, (byte) 0x72, (byte) 0x62, (byte) 0x01, (byte) 0x64, (byte) 0x44, (byte) 0x58,
				(byte) 0x73, (byte) 0x63, (byte) 0xe7, (byte) 0x22, (byte) 0x00, (byte) 0x76, (byte) 0x09, (byte) 0x69,
				(byte) 0x08, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x62,
				(byte) 0x06, (byte) 0x62, (byte) 0x00, (byte) 0x72, (byte) 0x63, (byte) 0x02, (byte) 0x01, (byte) 0x71,
				(byte) 0x01, (byte) 0x63, (byte) 0x35, (byte) 0x43, (byte) 0x00, (byte) 0x00 };
		DataInputStream is = new DataInputStream(new ByteArrayInputStream(message));

		SML_Message smlMessage = new SML_Message();
		Assert.assertTrue(smlMessage.decodeAndCheck(is));

		smlMessage = new SML_Message();
		Assert.assertTrue(smlMessage.decodeAndCheck(is));

		SML_GetListRes getListRes = (SML_GetListRes) smlMessage.getMessageBody().getChoice();
		for (SML_ListEntry listEntry : getListRes.getValList().getValListEntry()) {
			if (listEntry.getValTime().isSelected()) {
				if (listEntry.getValTime().getTag().getVal() == 3) {
					// System.out.println("timestamp: "
					// + ((SML_TimestampLocal)
					// (listEntry.getValTime().getChoice())).getTimestamp().getVal());
				}
			}
		}

		SML_Message smlMessage3 = new SML_Message();
		Assert.assertTrue(smlMessage3.decodeAndCheck(is));
		// System.out.println(Integer.toHexString(smlMessage3.getMessageBody().getTag().getVal()));

	}

}


src/sample/java/SampleSMLClient.java
/*
 * Copyright 2009-14 Fraunhofer ISE
 *
 * This file is part of jSML.
 * For more information visit http://www.openmuc.org
 *
 * jSML is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 2.1 of the License, or
 * (at your option) any later version.
 *
 * jSML is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with jSML.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

import java.io.ByteArrayOutputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.List;

import javax.net.ssl.SSLSocketFactory;

import org.openmuc.jsml.structures.OctetString;
import org.openmuc.jsml.structures.SML_File;
import org.openmuc.jsml.structures.SML_GetProfileListReq;
import org.openmuc.jsml.structures.SML_Message;
import org.openmuc.jsml.structures.SML_MessageBody;
import org.openmuc.jsml.structures.SML_PublicCloseReq;
import org.openmuc.jsml.structures.SML_PublicOpenReq;
import org.openmuc.jsml.structures.SML_TreePath;
import org.openmuc.jsml.structures.Unsigned8;
import org.openmuc.jsml.tl.SML_TConnection;
import org.openmuc.jsml.tl.SML_TSAP;

public class SampleSMLClient {

	public static void main(String[] args) throws UnknownHostException, IOException {

		if (args.length == 0) {
			System.out.println("Usage: SampleSMLClient portnum [-s]");
			System.exit(1);
		}

		int port = -1;
		try {
			port = Integer.parseInt(args[0]);
		} catch (NumberFormatException e) {
			System.out.println("Invalid port.");
			System.out.println("Usage: SampleSMLClient portnum [-s]");
			System.exit(1);
		}

		boolean useSSL = false;

		if (args.length > 1) {
			if (args[1].equals("-s")) {
				useSSL = true;
			}
			else {
				System.out.println("Usage: SampleSMLClient portnum [-s]");
				System.exit(1);
			}
		}

		SML_TSAP sml_tSAP = null;
		if (useSSL == false) {
			sml_tSAP = new SML_TSAP();
		}
		else {

			System.out.println("using ssl");

			// System.setProperty("javax.net.debug", "ssl");
			XTrustProvider.install();
			SSLSocketFactory sslSocketFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
			sml_tSAP = new SML_TSAP(sslSocketFactory);
		}

		SML_TConnection sml_tConnection = null;

		System.out.println("Trying to connect to port: " + port);
		sml_tConnection = sml_tSAP.connectTo(InetAddress.getByName("localhost"), port, 0);

		System.out.println("Sending a GetProfileListRequest");
		sml_tConnection.send(createGetProfileListRequest(1, 1));

		System.out.println("Listening for an SML response from the server");
		SML_File sml_File = sml_tConnection.receive();

		System.out.println("Got a response from the server.");

		List<SML_Message> sml_messages = sml_File.getMessages();
		for (SML_Message sml_message : sml_messages) {
			int tag = sml_message.getMessageBody().getTag().getVal();
			switch (tag) {
			case SML_MessageBody.OpenRequest:
				SML_PublicOpenReq sml_PublicOpenReq = (SML_PublicOpenReq) sml_message.getMessageBody().getChoice();
				System.out.println("Got OpenRequest");
				System.out.println("ClientID: " + sml_PublicOpenReq.getClientId().toString());
				break;
			case SML_MessageBody.OpenResponse:
				System.out.println("Got OpenResponse");
				break;
			case SML_MessageBody.CloseRequest:
				System.out.println("Got CloseRequest");
				break;
			case SML_MessageBody.CloseResponse:
				System.out.println("Got CloseResponse");
				break;
			case SML_MessageBody.GetProfilePackRequest:
				System.out.println("Got GetProfilePackRequest");
				break;
			case SML_MessageBody.GetProfilePackResponse:
				System.out.println("Got GetProfilePackResponse");
				break;
			case SML_MessageBody.GetProfileListRequest:
				System.out.println("Got GetProfileListRequest");
				break;
			case SML_MessageBody.GetProfileListResponse:
				System.out.println("Got GetProfileListResponse");
				break;
			case SML_MessageBody.GetProcParameterRequest:
				System.out.println("Got GetProcParameterRequest");
				break;
			case SML_MessageBody.GetProcParameterResponse:
				System.out.println("Got GetProcParameterResponse");
				break;
			case SML_MessageBody.SetProcParameterRequest:
				System.out.println("Got SetProcParameterRequest");
				break;
			case SML_MessageBody.GetListRequest:
				System.out.println("Got GetListRequest");
				break;
			case SML_MessageBody.GetListResponse:
				System.out.println("Got GetListResponse");
				break;
			case SML_MessageBody.AttentionResponse:
				System.out.println("Got AttentionResponse");
				break;
			default:
				System.out.println("type not found");
			}
		}

		System.out.println("Disconnecting");
		sml_tConnection.disconnect();

		System.out.println("Client is done and quits");
	}

	/**
	 * This sample builds a SML-File containing a PublicOpenRequest, a GetProfileListRequest and a PublicCloseRequest
	 * 
	 * @return
	 */
	public static byte[] createGetProfileListRequest(int groupNumber, int transactionIDCounter) {

		// The size of the ByteArray for the ByteArrayOutputStream should be
		// high enough so that expensive resizing is not needed. The
		// ByteArrayOutputStream can be reused by using the reset() function.
		ByteArrayOutputStream bs = new ByteArrayOutputStream(50);
		DataOutputStream os = new DataOutputStream(bs);

		// --- create a SML_Message with SML_PublicOpenRequest to start the
		// SML_File
		// make the transacionId and increment the transactionIDCounter
		OctetString transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		// make the group number and increment the group number
		Unsigned8 groupNum = new Unsigned8(groupNumber++);
		// continue on error
		Unsigned8 abortOnError = new Unsigned8(0x00);
		// create clientId
		int iClientId = 0x10;
		OctetString clientId = new OctetString(("" + iClientId).getBytes());
		// create reqFileId
		int iReqFileId = 0x11;
		OctetString reqFileId = new OctetString(("" + iReqFileId).getBytes());
		// create the SML_Message
		SML_PublicOpenReq openRequest = new SML_PublicOpenReq(null, clientId, reqFileId, null, null, null, null);
		SML_MessageBody smlMessageBody = new SML_MessageBody(SML_MessageBody.OpenRequest, openRequest);
		SML_Message openRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// --- create a SML_Message with SML_GetProfileListReq with test values
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		// make new group number
		groupNum = new Unsigned8(groupNumber++);
		SML_TreePath treePath = new SML_TreePath(new OctetString[] { new OctetString("testtreepath".getBytes()) });
		SML_GetProfileListReq getProfileListRequest = new SML_GetProfileListReq(null, null, null, null, null, null,
				treePath, null, null);
		smlMessageBody = new SML_MessageBody(SML_MessageBody.GetProfileListRequest, getProfileListRequest);
		SML_Message getProfileListRequestMessage = new SML_Message(transactionId, groupNum, abortOnError,
				smlMessageBody);

		// --- create a SML_Message with SML_PublicCloseRequest to end the
		// SML_File
		// make new transactionId
		transactionId = new OctetString(("" + transactionIDCounter++).getBytes());
		// make new group number
		groupNum = new Unsigned8(groupNumber++);
		SML_PublicCloseReq closeRequest = new SML_PublicCloseReq(null);
		smlMessageBody = new SML_MessageBody(SML_MessageBody.CloseRequest, closeRequest);
		SML_Message closeRequestMessage = new SML_Message(transactionId, groupNum, abortOnError, smlMessageBody);

		// encode every SML_Messages and write the encoded data to the stream os
		try {
			openRequestMessage.code(os);
			getProfileListRequestMessage.code(os);
			closeRequestMessage.code(os);
		} catch (IOException e) {
			e.printStackTrace();
			System.exit(1);
		}

		return bs.toByteArray();
	}

}
